### PR TITLE
[#224] Prefill when doing a `switch-to` or primary change.

### PR DESCRIPTION
**Pull request by fluca1978**
Whenever the primary host is changed, by means of an explicit
`switch-to` command or by a primary failure, the connection flushing
is activated. If possible, the prefill should be also restored on the
new server.
As suggested in
<https://github.com/agroal/pgagroal/pull/225#issuecomment-1117218107>
it would be nice to check if the specified new server is the same as
the old one (failure of the primary) or a different one (`switch-to`),
and in the case they are different the prefill is forced.

### CODE REVIEW

**Comment by jesperpedersen** [comment]
I think I like this direction a bit more.

Called the variable `primary` - and set it to `-1`, since we don't know what `server` really is. And, do the `server != (unsigned char)primary` check before the other prefill checks are done.

**Comment by jesperpedersen** [comment]
I would out in an explicit `primary != -1` check as well...

**Comment by jesperpedersen** [comment]
You don't need to enter the https://github.com/agroal/pgagroal/blob/master/src/libpgagroal/pool.c#L848 statement if primary isn't valid, so

```
int primary = -1;

if (pgagroal_get_primary(&primary))
{
    pgagroal_log_debug("No primary defined");
}

if (primary != -1 && server != (unsigned char)primary)
{
  // Line 848 and onwards ...

}

```

**Comment by jesperpedersen** [comment]
Can you squash and force push ?

**Comment by jesperpedersen** [comment]
Ah, there is still a thing.

We don't need to call prefill(false) for an old flushed primary server... which is the "if (primary != -1 && ..." check.

Otherwise it would try and prefill to `MIN_SIZE`

**Comment by fluca1978** [comment]
Just to make it clear, you don't want to prefill _at all_ if the primary is not set or the server has remained the same, right?
That makes me wondering: when it is fine to prefill to the MIN_VALUE?

**Comment by fluca1978** [comment]
Just to make it clear, you don't want to prefill _at all_ if the primary is not set or the server has remained the same, right?
That makes me wondering: when it is fine to prefill to the MIN_VALUE?

**Comment by jesperpedersen** [comment]
This is only for the `flush_server` case, and yes - there should be no prefill (as `MIN_SIZE` could be > 0, and therefore cause connects to an invalid host).

There are a lot of cases where `prefill(false)` makes sense - f.ex. after validation of a connection fails or after idle timeout. This is why people need to consider `MIN_SIZE` carefully as it will tie up resources on the database cluster.

**Comment by fluca1978** [comment]
I'm not really comfortable with squashing a set of published commits, please check if it is fine now.

File: src/libpgagroal/pool.c [hunk 1]
```diff
@@ -845,11 +847,21 @@ pgagroal_flush_server(signed char server)
       }
    }
 
-   if (config->number_of_users > 0 && config->number_of_limits > 0)
+   if (pgagroal_get_primary(&primary))
+      pgagroal_log_debug("No primary defined");
+   else
    {
-      if (!fork())
+      prefill = server != (unsigned char)primary && primary != -1;
```
Comment (jesperpedersen) [inline src/libpgagroal/pool.c:854]: This needs to go into the L856 if check - and therefore you can remove  the `prefill` variable

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
And, you can just do 2 levels

```
if (primary != -1 && ...
{
   // Existing code
  if (..
```

**Comment by fluca1978** [comment]
Sorry, I'm a little confused: if there is the need to prefill (number of users and limits not zero) we `fork` and issue a `pgagroal_prefill` in the child process, that then exits, while the parent process continues to memory destroy and exits too.
So what is the need to `fork` if not issuing a `pgagroal_prefill`?

**Comment by jesperpedersen** [comment]
I have to check something with uncrustify - otherwise I'll merge :)

**Comment by jesperpedersen** [comment]
Yeah, it could be a `static` function within `pool.c`

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Pull request closed without merging**